### PR TITLE
Implement Update method for EndpointManager

### PR DESCRIPTION
**Pull request by lemonlinger**
This is a part of changes for https://github.com/etcd-io/etcd/issues/12652 including:
- Add testing for the resolver.
- Implement Update method for EndpointManager.

@peterbourgon

### CODE REVIEW

File: client/v3/naming/endpoints/endpoints_impl.go [hunk 1]
```diff
@@ -4,37 +4,64 @@ package endpoints
...
-	//	}
-	//	_, err = gr.Client.KV.Put(ctx, target+"/"+nm.Addr, string(v), opts...)
-	//case internal.Delete:
-	//	_, err = gr.Client.Delete(ctx, target+"/"+nm.Addr, opts...)
-	//default:
-	//	return status.Error(codes.InvalidArgument, "naming: bad naming op")
+func (m *endpointManager) Update(ctx context.Context, updates []*UpdateWithOpts) (err error) {
+	ops := make([]clientv3.Op, 0, len(updates))
+	for _, update := range updates {
+		switch update.Op {
```
Comment (ptabor) [inline client/v3/naming/endpoints/endpoints_impl.go:44]: We should verify probably here whether `target`+"/" is prefix of `key`.

File: tests/integration/clientv3/naming/resolver_test.go [hunk 2]
```diff
@@ -15,56 +15,103 @@
...
 
 func TestEtcdGrpcResolver(t *testing.T) {
-	t.Skip("Not implemented yet")
-
 	defer testutil.AfterTest(t)
 
-	// s1 :=  // TODO: Dummy GRPC service listening on 127.0.0.1:20000
-	// s2 :=  // TODO: Dummy GRPC service listening on 127.0.0.1:20001
+	s1PayloadBody := []byte{'1'}
+	s1 := &integration.StubServer{
```
Comment (ptabor) [inline tests/integration/clientv3/naming/resolver_test.go:38]: Probably worth to write helper method for s1/s2 creation as the logic is shared.

File: tests/integration/clientv3/naming/resolver_test.go [hunk 3]
```diff
@@ -15,56 +15,103 @@
...
 
 	em.DeleteEndpoint(context.TODO(), "foo/e1")
 	em.AddEndpoint(context.TODO(), "foo/e2", e2)
 
-	// TODO: Send requests to conn and make sure s2 receive it.
-	// Might require restarting s1 to break the existing (open) connection.
-
-	conn.GetState() // this line is to avoid compiler warning that conn is unused.
+	// Give gRPC Client a while to update underlying connections since it's asynchronous.
+	time.Sleep(100 * time.Millisecond)
```
Comment (ptabor) [inline tests/integration/clientv3/naming/resolver_test.go:109]: Please make it in a loop with deadline of ~30s. 
I don't want to risk test getting flake if it takes 500ms or even 2s...  at the same time I want to detect that the condition happend quickly, so in the loop we should sleep this order of ~100ms.

File: tests/integration/grpc_stub_server.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+// StubServer is borrowed from the interal package of grpc-go.
+// See https://github.com/grpc/grpc-go/blob/master/internal/stubserver/stubserver.go
+// Since it cannot be imported directly, we have to copy and paste it here,
+// and useless code for our testing is removed.
+
+// StubServer is a server that is easy to customize within individual test
+// cases.
+type StubServer struct {
```
Comment (ptabor) [inline tests/integration/grpc_stub_server.go:19]: Nice.

I would consider putting this into: 
     `etcd/pkg/testutil`
or  even  `etcd/pkg/grpc_testing`(slightly preferred)

as this is etcd-independent and fulfills criteria for pkg.

File: client/v3/naming/endpoints/endpoints_impl.go [hunk 5]
```diff
@@ -116,6 +143,19 @@ func (m *endpointManager) NewWatchChannel(ctx context.Context) (WatchChannel, er
 }
 
 func (m *endpointManager) List(ctx context.Context) (Key2EndpointMap, error) {
```
Comment (ptabor) [inline client/v3/naming/endpoints/endpoints_impl.go:145]: Thank you for implementing this. 

So if we have a complementary methods (Add, Delete, Update, List), 
could you, please, add integration test for endpoints alone (just that DB operations behave as expected).

**Review by ptabor — CHANGES_REQUESTED** [review]
Thank you. That looks awesome. 
Just a few 'feature_requests' in the comments.

File: tests/integration/clientv3/naming/resolver_test.go (hunk 2 above)
Comment (lemonlinger) [reply tests/integration/clientv3/naming/resolver_test.go]: Done

File: tests/integration/clientv3/naming/resolver_test.go (hunk 3 above)
Comment (lemonlinger) [reply tests/integration/clientv3/naming/resolver_test.go]: Done

File: tests/integration/grpc_stub_server.go (hunk 4 above)
Comment (lemonlinger) [reply tests/integration/grpc_stub_server.go]: Done

File: client/v3/naming/endpoints/endpoints_impl.go (hunk 5 above)
Comment (lemonlinger) [reply client/v3/naming/endpoints/endpoints_impl.go:150]: Done

File: client/v3/naming/endpoints/endpoints_impl.go (hunk 1 above)
Comment (lemonlinger) [reply client/v3/naming/endpoints/endpoints_impl.go:49]: Done

**Review by ptabor — APPROVED** [review]
Thank you.

**Comment by ptabor** [comment]
PASSES="fmt" ./test.sh

Failed with: 
```
go: updates to go.sum needed, disabled by -mod=readonly
FAIL: 'go_fmt_for_package ./...' checking failed (!=0 return code)
```

Please run ./scripts/fix.sh

**Comment by ptabor** [comment]
Still 
```
go: updates to go.sum needed, disabled by -mod=readonly
must be run from 'go.etcd.io/etcd/v3' module directory
```

Please check ```go mod tidy``` in the main directory if the ./scripts/fix.sh does not help.

**Comment by lemonlinger** [comment]
> Still
> 
> ```
> go: updates to go.sum needed, disabled by -mod=readonly
> must be run from 'go.etcd.io/etcd/v3' module directory
> ```
> 
> Please check `go mod tidy` in the main directory if the ./scripts/fix.sh does not help.

Run `go mod tidy` but nothing happened.

**Comment by ptabor** [comment]
Assuming it passes, please squash the commits.

**Comment by ptabor** [comment]
Woohu. Don't worry about the flake. Please squash and I will merge.

**Comment by lemonlinger** [comment]
> Woohu. Don't worry about the flake. Please squash and I will merge.

Sorry for the slow response. Squashed.

**Comment by ptabor** [comment]
> Sorry for the slow response. Squashed.
Your response time is impressive. Thank you.

Merging on green.

**Pull request merged**